uvicorn/protocols/http/httptools_impl.py

```diff
@@ -151,6 +151,8 @@ def connection_lost(self, exc):
             self.cycle.message_event.set()
         if self.flow is not None:
             self.flow.resume_writing()
+        if exc is None:
+            self.transport.close()
```

Comment:
Could you explain these two lines?

The questions I have are:

- I assume this was prompted by warnings raising errors in certain situations. Did it happen always or only in specific conditions?
- Maybe related, why only close in case there was no exception? Does closing in case of exception cause issues?

Reply:
ok if you remove those 2 lines that close the transport inside the connection_lost method you get errors in the `test_ssl`

```
E               Traceback (most recent call last):
E                 File "/home/lotso/.asdf/installs/python/3.9.1/lib/python3.9/asyncio/sslproto.py", line 320, in __del__
E                   _warn(f"unclosed transport {self!r}", ResourceWarning, source=self)
E               ResourceWarning: unclosed transport <asyncio.sslproto._SSLProtocolTransport object at 0x7f3d548c8460>

../venv/lib/python3.9/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning
```

it's only true for `httptools_impl.py` you dont have the error should you remove the exact same lines in `h11_impl.py` or the 2 ws protocols.

reason it works without for those last 3 is simply (I think) the ssl tests defaults to the httptools implementation, so I decided to put it in 4 places but it's a bet.

the `ResourceWarning` error happens on every test, it's very reproducible.

the reason I opted for the `exc` condition is debatable, I'm not sure about it : I saw in the `BaseProtocol` the following docstring:

```
    def connection_lost(self, exc):
        """Called when the connection is lost or closed.

        The argument is an exception object or None (the latter
        meaning a regular EOF is received or the connection was
        aborted or closed).
        """
```

while debugging it I had a None parameter, so I decided to close the transport only in that case, but I'll be lying if I said I'm 100% sure it's the correct move

Reply:
I'd be also interested in knowing if that change (closing the transport to fix the test_ssl ressource warning) would help on the weird issue https://github.com/encode/uvicorn/issues/1021